r-package/grf/vignettes/llf.Rmd

```diff
@@ -0,0 +1,147 @@
+
+```{r}
+# Select covariates 
+lasso.mod <- cv.glmnet(X, Y, alpha = 1)
+lasso.coef <- predict(lasso.mod, type = "nonzero")
+selected <- ifelse(is.null(dim(lasso.coef)), 1:ncol(X), lasso.coef[,1])
+
+# Train LLF
+llf.lasso.preds <- predict(forest, X.test, 
+                           linear.correction.variables = selected,
+                           ll.weight.penalty = TRUE)
+results.llf.lasso <- llf.lasso.preds$predictions
+mean((results.llf.lasso - truth)**2)
+```
+
+Last, consider variance estimates and confidence intervals. A user must specify estimate.variance to be TRUE. A parameter under the hood is ci.group.size, which must be at least 2 in order to grow variance estimates (defaults to 2); the forest will grow ci.group.size trees on every subsample in order to provide confidence intervals. We use a different data-generating process for easier visualization.
+
+```{r}
+mu = function(x){ log(1 + exp(6 * x)) }
+
+X = matrix(runif(n*p, -1, 1), nrow = n)
+Y = mu(X[,1]) + sigma * rnorm(n)
+
+X.test = matrix(runif(n*p, -1, 1), nrow = n)
+ticks = seq(-1, 1, length = n)
+X.test[,1] = ticks
+truth = mu(ticks)
+
+ll.forest <- ll_regression_forest(X, Y, enable.ll.split = TRUE)
+results.llf.var <- predict(ll.forest, X.test, 
+                           linear.correction.variables = selected, 
+                          estimate.variance = TRUE)
+preds.llf.var <- results.llf.var$predictions
+variance.estimates <- results.llf.var$variance.estimates
+
+# find lower and upper bounds for 95% intervals 
+lower.llf <- preds.llf.var - 1.96*sqrt(variance.estimates)
+upper.llf <- preds.llf.var + 1.96*sqrt(variance.estimates)
+
+df <- data.frame(cbind(ticks, truth, preds.llf.var, lower.llf, upper.llf))
+ggplot(df, aes(ticks)) + 
+  geom_point(aes(y = preds.llf.var, color = "Local Linear Forest"), show.legend = F, size = 0.6) +
+  geom_line(aes(y = truth)) + 
+  geom_line(aes(y = lower.llf), color = "gray", lty = 2) + 
+  geom_line(aes(y = upper.llf), color = "gray", lty = 2) + 
+  xlab("x") + ylab("y") + theme_bw()
+```
+
+
+A note on larger datasets: Local linear forests work well in low dimensions; when we get more data, however, it starts to take a very long time to run. This is because with n_train and n_test train and test points, we are running n_test regressions with n_train data points each. However, sometimes we still want to use random forests and correct for linear trends. In this case (datasets with roughly 100,000 or more observations, although always context-dependent), selecting a small number linear correction variables is especially important. The current ridge parameter tuning will also take prohibitively long, and so we recommend either setting the value to 0.01 consistently, tuning this on a subset of the data, restricting the range of values considered, or cross-validating using a small number of shallow trees.  
```

Comment:
Consider moving this closer to the section where you talk about variable selection and the lasso?